src/auth0-session/client/edge-client.ts

```diff
@@ -0,0 +1,241 @@
+import * as oauth from 'oauth4webapi';
+import * as jose from 'jose';
+import { Auth0Request } from '../http';
+import {
+  CallbackExtras,
+  OpenIDCallbackChecks,
+  TokenEndpointResponse,
+  AbstractClient,
+  EndSessionParameters,
+  Telemetry
+} from './abstract-client';
+import { ApplicationError, DiscoveryError, IdentityProviderError } from '../utils/errors';
+import { AccessTokenError, AccessTokenErrorCode } from '../../utils/errors';
+import urlJoin from 'url-join';
+import { Config } from '../config';
+
+const encodeBase64 = (input: string) => {
+  const unencoded = new TextEncoder().encode(input);
+  const CHUNK_SIZE = 0x8000;
+  const arr = [];
+  for (let i = 0; i < unencoded.length; i += CHUNK_SIZE) {
+    // @ts-expect-error Argument of type 'Uint8Array' is not assignable to parameter of type 'number[]'.
+    arr.push(String.fromCharCode.apply(null, unencoded.subarray(i, i + CHUNK_SIZE)));
+  }
+  return btoa(arr.join(''));
+};
+
+export class EdgeClient extends AbstractClient {
+  private client?: oauth.Client;
+  private as?: oauth.AuthorizationServer;
+  private httpOptions: () => oauth.HttpRequestOptions;
+
+  constructor(protected config: Config, protected telemetry: Telemetry) {
+    super(config, telemetry);
+    if (config.authorizationParams.response_type !== 'code') {
+      throw new Error('This SDK only supports `response_type=code` when used in an Edge runtime.');
+    }
+
+    this.httpOptions = () => {
+      const headers = new Headers();
+      if (config.enableTelemetry) {
+        const { name, version } = telemetry;
+        headers.set('User-Agent', `${name}/${version}`);
+        headers.set(
+          'Auth0-Client',
+          encodeBase64(
+            JSON.stringify({
+              name,
+              version,
+              env: {
+                edge: true
+              }
+            })
+          )
+        );
+      }
+      return {
+        signal: AbortSignal.timeout(this.config.httpTimeout),
+        headers
+      };
+    };
+  }
+
+  private async getClient(): Promise<[oauth.AuthorizationServer, oauth.Client]> {
+    if (this.as) {
+      return [this.as, this.client as oauth.Client];
+    }
```

Comment:
Not sure how long-lived instances in the Edge runtime can be (apparently [there's still some notion](https://news.ycombinator.com/item?id=31912794) of "cold-start"). If these were short-lived, wouldn't we be calling the discovery endpoint a lot?

Reply:
We can't do any better than keeping them in memory 🤷 (unless Next.js has some built in distributed cache I haven't heard of)

Reply:
Could we have an option to not use the discovery endpoint?

Reply:
The question of in memory discovery responses not being persisted is not a new thing we've introduced now that it's running on the Edge runtime. Serverless functions that serve the Node runtime also have the same behaviour, so this SDK has had this behaviour since 0.1.

We could ask users to provide enough configuration to skip the Discovery step, sure (we could also do this in express) - but I don't think this is a feature we need to add just because the client can now be served from Edge Serverless functions as well as Node Serverless functions.

Reply:
Asking because serverless !== edge. As Vercel's architect mentions, "Edge Functions bring another set of trade-offs into the mix".

Reply:
What was enough/serviceable with serverless might not be with edge functions.

Reply:
In any case, this probably warrants more investigation and testing with the edge runtime to find out if something more needs to be done.

Reply:
Since edge functions apparently use Cloudflare Workers, I guess it comes down to the differences between Cloudflare Workers and AWS Lambda.

Reply:
This SDK is agnostic to a particular edge vendor (Vercel/AWS/Azure/CloudFlare/GCP) - but sure, we can look into how their memory persistence differs between the 2 runtimes. I guess at the very least, Edge functions will be more populous given that their are more Edge locations - so I guess the probability of hitting the same box is less. But on the other hand, the Discovery is served from a CDN, so it's not a huge cost.